docs/writing-rules/glossary.md

```diff
@@ -0,0 +1,124 @@
+tags:
+  - Rule writing
+---
+
+# Static analysis and rule-writing glossary
+
+The definitions provided here are specific to each term's meaning and use in Semgrep.
+
+## Constant propagation
+
+Refers to state of a variable remaining constant throughout the program. Semgrep can analyze whether a variable carries a constant value at a given point. Both Semgrep OSS and Semgrep Pro Engine perform this analysis, with Semgrep Pro able to track the propagation across files.
+
+You can write checks for constant propagation through the use of `pattern-not` in conjunction with `pattern` or `metavariable-comparison`.
+
+It is enabled by default when writing rules and [can be disabled](/data-flow/constant-propagation/#disabling-constant-propagation).
+
+## Cross-file analysis
+
+Also known as **interfile analysis**. This analysis traces or tracks data and its transformations across files, such as when a variable is defined in one file but used in another.
+
+Cross file analysis can be paired with taint analysis to detect unsanitized variables flowing from a source to a sink.
+
+Also known as **interfile analysis**. Available in Semgrep Pro Engine.
+
+## Cross-function analysis
+
+This analysis traces or tracks data and its transformations across functions in a single file, such as when a globally-scoped variable is defined in one function but used in another.
+
+Also known as **intrafile** or **interprocedural** analysis. Available in Semgrep Pro Engine.
+
+## Error matrix
+
+An error matrix is a 2x2 table that visualizes the performance of a Semgrep rule in relation to the findings it does or doesn't detect. It has two axes:
+
+- Positive and negative
+- True or false
+
+These yield the following combinations:
+
+<dl>
+<dt>True positive</dt>
+<dd>The rule detected a piece of code it was intended to find.</dd>
+<dt>False positive</dt>
+<dd>The rule detected a piece of code it was not intended to find.</dd>
+<dt>True negative</dt>
+<dd>The rule correctly skipped over a piece of code it wasn't meant to find.</dd>
+<dt>False negative</dt>
+<dd>The rule failed to detect a piece of code it should have found.</dd>
+</dl>
+
+Not to be confused with **risk matrices**.
+
+## Finding
+
+A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. Findings can be security issues, bugs, or code that doesn't follow coding conventions.
+
+## Fully-qualified name
+
+A fully-qualified name refers to a name which uniquely identifies a class, method, type, or module. Languages such as C# and Ruby use `::` to distinguish between fully-qualified names and regular names.
+
+Not to be confused with **tokens**.
+
+## l-value (left-, or location-value)
+
+An expression that denotes an object in memory; a memory location, something that you can use in the left-hand side (LHS) of an assignment.
+
+Compare to [r-value](https://learn.microsoft.com/en-us/cpp/c-language/l-value-and-r-value-expressions?view=msvc-170).
```

Comment:
I don't think we need to compare this to r-value! The main thing is that it can be used in the left hand side of an assignment. I'd also lean against including the microsoft docs, which might give the impression that l-values are only a C/C++ concept.

If we want to clarify this more, we could give some examples. `x` and `arr[2]` are l-values (you could write `x = 122` or `arr[2] = "semgrep"`. However, `2 + 2` is not an l-value (you could not write `2 + 2 = 122`).